lib/puppetserver/ca/action/prune.rb

```diff
@@ -0,0 +1,130 @@
+require 'benchmark'
+require 'optparse'
+require 'openssl'
+require 'puppetserver/ca/errors'
+require 'puppetserver/ca/utils/cli_parsing'
+require 'puppetserver/ca/utils/file_system'
+require 'puppetserver/ca/utils/config'
+require 'puppetserver/ca/x509_loader'
+
+module Puppetserver
+  module Ca
+    module Action
+      class Prune
+        include Puppetserver::Ca::Utils
+
+        SUMMARY = "Prune the local CRL on disk to remove any duplicated certificates"
+        BANNER = <<-BANNER
+Usage:
+  puppetserver ca prune [--help]
+  puppetserver ca prune [--config]
+
+Description:
+  Prune the list of revoked certificates of any duplication within it.
+
+Options:
+BANNER
+
+        def initialize(logger)
+          @logger = logger
+        end
+
+        def self.parser(parsed = {})
+          OptionParser.new do |opts|
+            opts.banner = BANNER
+            opts.on('--help', 'Display this command-specific help output') do |help|
+              parsed['help'] = true
+            end
+            opts.on('--config CONF', 'Path to the puppet.conf file on disk') do |conf|
+              parsed['config'] = conf
+            end
+          end
+        end
+
+        def run(inputs)
+          config_path = inputs['config']
+
+          # Validate the config path.
+          if config_path
+            errors = FileSystem.validate_file_paths(config_path)
+            return 1 if Errors.handle_with_usage(@logger, errors)
+          end
+
+          # Validate puppet config setting.
+          settings_overrides = {}
+          puppet = Config::Puppet.new(config_path)
+          puppet.load(settings_overrides)
+          return 1 if Errors.handle_with_usage(@logger, puppet.errors)
+
+          # Validate that we are offline
+          return 1 if HttpClient.check_server_online(puppet.settings, @logger)
+
+          # Getting the CRL(s)
+          loader = X509Loader.new(puppet.settings[:cacert], puppet.settings[:cakey], puppet.settings[:cacrl])
+
+          # Prune the CRL(s)
+          crl_list = loader.crls
+          prune_CRLs(crl_list)
+          write_crl_to_file(crl_list, puppet.settings[:cacrl])
+
+          @logger.inform("Finished pruning your local CRLs")
+          return 0
+        end
+
+        def prune_CRLs(crl_list)
+          crl_list.each do |list|
```

Comment:
So, thinking about this a bit more, we don't want to prune CRLs that we don't own. After having deduplicated a CRL we need to increment its version number and re-sign it before writing it to disk. To sign it we need the private key of the CRL issuer, which will only be for our leaf CA.

Reply:
So I should implement a check to see if a CRL's owner is us and then only prune those selected CRLs.

Reply:
yeah, I believe that should just be the first CRL from the `ca_crl.pem` file, but would probably be good to check. But I don't _think_ we need to loop through the rest of the CRLs once we have ours.

Reply:
My understand on CRL are a bit limited beside looking at Ruby classes.  Is `ca_crl.pem` contains only the CRLS that are ours?  I thought it was that way since it's in the puppetlabs directory on my local computer.

Reply:
There's a few different scenarios you can run Puppet Server's CA under:
1) We operate a single self-signed certificate for our CA which signs every other puppet cert. (Deprecated, no longer default, but still common with folks upgrading)
2) We create a long lived, self-signed root cert that is stored on the server but not used for anything but signing the shorter lived "child" or "leaf" cert. The root in this sense is "offline" and the leaf is "active". This allows a shorter CA cert for security purposes, you can revoke your existing cert and use the root to create a new one that your agent's will still trust. (Current default out of the box).
3) There's a corporate root CA cert which then has a chain of children, one of which being the Puppet CA. This is common in large enterprises.

So, it's possible for the CRL chain to one or more certs, and when there's only two certs we may "own" the root and have the private key on disk, but that root can be considered "offline" and doesn't need maintenance. The other common case when there's 2+ CRLs is that we only own the leaf CRL and should only act on it.

Sorry, that wasn't spelled out clearly in the ticket.

Reply:
No, the `ca_crl.pem` file contains a CRL for each certificate in the certificate chain. We allow people to use their own external root CA certificates, so there may be certs, and therefore corresponding CRLs, that we don't own.

This page has more detail, including what to put in each of the files: https://puppet.com/docs/puppet/7/server/intermediate_ca.html

Reply:
Interesting.  Based on that website from @mwaggett , the first CRL is probably all that I need then (still gonna check for their name tho).  Thanks for explaining it further guys!